applications/GeoMechanicsApplication/custom_strategies/schemes/geomechanics_time_integration_scheme.hpp

```diff
@@ -0,0 +1,360 @@
+
+        KRATOS_CATCH("")
+    }
+
+    void CalculateLHSContribution(Element& rCurrentElement,
+                                  LocalSystemMatrixType& LHS_Contribution,
+                                  Element::EquationIdVectorType& EquationId,
+                                  const ProcessInfo& CurrentProcessInfo) override
+    {
+        KRATOS_TRY
+
+        rCurrentElement.CalculateLeftHandSide(LHS_Contribution, CurrentProcessInfo);
+
+        rCurrentElement.EquationIdVector(EquationId, CurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    void CalculateLHSContribution(Condition& rCurrentCondition,
+                                  LocalSystemMatrixType& LHS_Contribution,
+                                  Element::EquationIdVectorType& EquationId,
+                                  const ProcessInfo& CurrentProcessInfo) override
+    {
+        KRATOS_TRY
+
+        rCurrentCondition.CalculateLeftHandSide(LHS_Contribution, CurrentProcessInfo);
+
+        rCurrentCondition.EquationIdVector(EquationId, CurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    void Update(ModelPart& rModelPart,
+                DofsArrayType& rDofSet,
+                TSystemMatrixType& A,
+                TSystemVectorType& Dx,
+                TSystemVectorType& b) override
+    {
+        KRATOS_TRY
+
+        int num_threads = ParallelUtilities::GetNumThreads();
+        OpenMPUtils::PartitionVector dof_set_partition;
+        OpenMPUtils::DivideInPartitions(static_cast<int>(rDofSet.size()),
+                                        num_threads, dof_set_partition);
+
+#pragma omp parallel
+        {
+            int k = OpenMPUtils::ThisThread();
+
+            typename DofsArrayType::iterator dofs_begin =
+                rDofSet.begin() + dof_set_partition[k];
+            typename DofsArrayType::iterator dofs_end =
+                rDofSet.begin() + dof_set_partition[k + 1];
+
+            // Update Displacement and Pressure (DOFs)
+            for (typename DofsArrayType::iterator itDof = dofs_begin;
+                 itDof != dofs_end; ++itDof) {
+                if (itDof->IsFree())
+                    itDof->GetSolutionStepValue() +=
+                        TSparseSpace::GetValue(Dx, itDof->EquationId());
+            }
+        }
+
+        this->UpdateVariablesDerivatives(rModelPart);
+
+        KRATOS_CATCH("")
+    }
+
+protected:
+    void CheckBufferSize(const ModelPart& rModelPart) const
+    {
+        constexpr int minimum_buffer_size = 2;
+        KRATOS_ERROR_IF(rModelPart.GetBufferSize() < minimum_buffer_size)
+            << "insufficient buffer size. Buffer size should be "
+               "greater or equal to "
+            << minimum_buffer_size << ". Current size is "
+            << rModelPart.GetBufferSize() << std::endl;
+    }
+
+    void CheckSolutionStepsData(const Node& r_node, const Variable<double>& variable) const
+    {
+        KRATOS_ERROR_IF(!r_node.SolutionStepsDataHas(variable))
```

Comment:
Perhaps it's slightly better to use the `..._NOT` macro here, since that is harder to overlook than a `!`:
```suggestion
        KRATOS_ERROR_IF_NOT(r_node.SolutionStepsDataHas(variable))
```

Reply:
Agreed, done